Quick fix for possible segfault in PurgeConnection

jasonmp85 opened the pull request:
Now that connections can be acquired without going through the cache, we have to handle cases where functions assume the cache has been initialized.

Fixes #664

anarazel commented:
Won't this mean that we'll issue
```
	/*
	 * It's possible the provided connection matches the host and port for
	 * an entry in the hash without being precisely the same connection. In
	 * that case, we will want to close the provided connection in addition
	 * to the one from the hash (which was closed by PurgeConnectionByKey).
	 */
	if (purgedConnection != connection)
	{
		ereport(WARNING, (errmsg("hash entry for \"%s:%d\" contained different "
								 "connection than that provided by caller",
								 nodeConnectionKey.nodeName,
								 nodeConnectionKey.nodePort)));
		PQfinish(connection);
	}
```
will be emitted regularly?  Comment should be updated as well.

jasonmp85 commented:
> Won't this mean that [this] will be emitted regularly?

Gah, yes. I'm just going to remove that. I don't think it's ever helped fix anything. In addition, I'll add a test when I get back.

jasonmp85 commented:
> Comment should be updated as well.

I removed the `ereport`, but I believe the comment remains accurate… It's still the case that someone can pass in a connection that's not in the cache (including the vacuous case where the cache doesn't exist), and it's still the case that our contract with the caller is that we'll close it. So leaving as-is.

I added a test, rebased to put it before my fix, and `git reset` to try it out. Segfaults as expected. The fix makes the segfault go away. Will merge to keep test history clean after :shipit: is received.

marcocitus commented on src/backend/distributed/utils/connection_cache.c:
```diff
@@ -186,11 +182,17 @@ PurgeConnectionByKey(NodeConnectionKey *nodeConnectionKey)
 
-	nodeConnectionEntry = hash_search(NodeConnectionHash, nodeConnectionKey, HASH_REMOVE,
-									  &entryFound);
 	if (entryFound)
```
We could add PQsocket comparison here to make sure we're actually talking about the same connection.

marcocitus commented on src/test/regress/expected/multi_connection_cache.out:
```diff
@@ -104,3 +108,12 @@ SELECT get_and_purge_connection('localhost', :worker_port);
+\c
+-- purge existing connection to localhost
+SELECT connect_and_purge_connection('localhost', :worker_port);
+WARNING:  could not find hash entry for connection to "localhost:57638"
```
It seems the user might see this message in some scenarios. Is it relevant?

jasonmp85 replied:
Arg! I thought I took this out; missed it in the test output. I see now that I have such a check in both `Purge` methods. Fixed.

anarazel commented on src/test/regress/expected/multi_connection_cache.out:
```diff
@@ -104,3 +108,11 @@ SELECT get_and_purge_connection('localhost', :worker_port);
 SET client_min_messages TO DEFAULT;
+\c
+-- purge existing connection to localhost
+SELECT connect_and_purge_connection('localhost', :worker_port);
```
I think doing this test last doesn't actually reproduce the issue at hand here, does it? Because the hash is already created...

jasonmp85 replied:
That's what the /c is there for, right? It reconnects, resetting the
session.

It definitely segfaults (tested it).

--
*Jason Petersen*
Software Engineer | Citus Data
303.736.9255
jason@citusdata.com

On Jul 28, 2016, at 5:53 PM, Andres Freund <notifications@github.com> wrote:

In src/test/regress/expected/multi_connection_cache.out
<https://github.com/citusdata/citus/pull/675#discussion_r72723448>:

> @@ -104,3 +108,11 @@ SELECT get_and_purge_connection('localhost', :worker_port);
>  (1 row)
>
>  SET client_min_messages TO DEFAULT;
> +\c
> +-- purge existing connection to localhost
> +SELECT connect_and_purge_connection('localhost', :worker_port);

I think doing this test last doesn't actually reproduce the issue at hand
here, does it? Because the hash is already created...

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/citusdata/citus/pull/675/files/c52365b51155347a6c01b2179b3ca2b701da0168#r72723448>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AADF8dAjF38sgtiKRM9tSEMoIG9qECWFks5qaUDxgaJpZM4JVeTE>
.

anarazel replied:
On 2016-07-28 17:01:42 -0700, Jason Petersen wrote:
> > @@ -104,3 +108,11 @@ SELECT get_and_purge_connection('localhost', :worker_port);
> >  (1 row)
> >  
> >  SET client_min_messages TO DEFAULT;
> > +\c
> > +-- purge existing connection to localhost
> > +SELECT connect_and_purge_connection('localhost', :worker_port);
> 
> That's what the /c is there for, right? It reconnects, resetting the
> session.

Err, yes. Sorry.

anarazel commented:
Besides the one test issue, I'm good with that. :shipit:
